runners/spark/src/main/java/org/apache/beam/runners/spark/util/SparkCompat.java

```diff
@@ -0,0 +1,241 @@
+   * we cannot ensure this.
+   *
+   * <p>This is an internal class used to ensure compatibility between spark versions and should not
+   * be used for other purposes.
+   */
+  public static class IterableIterator<T> implements Iterable<T>, Iterator<T> {
+
+    private final Iterable<T> delegate;
+    private final Iterator<T> iter;
+
+    public IterableIterator(Iterable<T> delegate) {
+      this.delegate = delegate;
+      this.iter = iterator();
+    }
+
+    public IterableIterator(final Iterator<T> iter) {
+      this.delegate =
+          new Iterable<T>() {
+            @Override
+            public Iterator<T> iterator() {
+              return iter;
+            }
+          };
+      this.iter = iter;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+      // This is done explicitly to respect the semantics of Iterable.
+      // In case someone wants to ask for new iterators from an wrapped Iterable more than once.
+      // However if we are wrapping an Iterator it won't produce a new Iterator but its reference.
+      return delegate.iterator();
+    }
+
+    @Override
+    public boolean hasNext() {
+      return iter.hasNext();
+    }
+
+    @Override
+    public T next() {
+      return iter.next();
+    }
+
+    @Override
+    public void remove() {
+      iter.remove();
+    }
+  }
+
+  /**
+   * A SparkCompat.IterableIterator compatible version of {@link
+   * org.apache.spark.api.java.function.FlatMapFunction}.
+   */
+  public interface FlatMapFunction<K, V>
+      extends org.apache.spark.api.java.function.FlatMapFunction<K, V> {
+
+    @Override
+    SparkCompat.IterableIterator<V> call(K k) throws Exception;
+  }
+
+  /**
+   * A SparkCompat.IterableIterator compatible version of {@link
+   * org.apache.spark.api.java.function.PairFlatMapFunction}.
+   */
+  public interface PairFlatMapFunction<T, K, V>
+      extends org.apache.spark.api.java.function.PairFlatMapFunction<T, K, V> {
+
+    @Override
+    SparkCompat.IterableIterator<Tuple2<K, V>> call(T t) throws Exception;
+  }
+
+  /**
+   * A spark 1/2 compatible version of the Spark's {@link org.apache.spark.storage.BlockManager}.
+   */
+  abstract static class BlockManager {
+    static final BlockId WATERMARKS_BLOCK_ID = BlockId.apply("broadcast_0WATERMARKS");
+
+    void removeBlock(BlockId blockId, boolean tellMaster) {
+      final SparkEnv sparkEnv = SparkEnv.get();
+      if (sparkEnv != null) {
+        final org.apache.spark.storage.BlockManager blockManager = sparkEnv.blockManager();
+        blockManager.removeBlock(WATERMARKS_BLOCK_ID, true);
+      }
+    }
+
+    abstract void putSingle(Map<Integer, GlobalWatermarkHolder.SparkWatermarks> empty);
+
+    abstract Option<BlockResult> getRemote();
+  }
+
+  /**
+   * A Factory Method that detects the version of Spark and creates and appropriate BlockManager
+   * instance.
+   */
+  static BlockManager newBlockManager() {
```

Comment:
Perhaps this method could be renamed to `blockManager()`, similarly to `SparkEnv.get().blockManager()`.

While it is guaranteed that the caller gets a new instance of the wrapper (i.e. `Spark2BlockManager` / `Spark1BlockManager`), the underlying instance of the actual `BlockManager` will remain the same since `SparkEnv.get().blockManager()` essentially returns a field (`blockManager` is a `val` in `SparkEnv`).